src/opnsense/scripts/unbound/logger.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/local/bin/python3
+
+import sys
+import selectors
+import argparse
+import syslog
+import sqlite3
+import time
+import datetime
+from timeit import default_timer as timer
+from collections import deque
+sys.path.insert(0, "/usr/local/opnsense/site-python")
+from daemonize import Daemonize
+from sqlite3_helper import check_and_repair
+
+class DNSReader:
+    def __init__(self, flush_interval):
+        check_and_repair("/var/unbound/data/unbound.sqlite")
+
+        self.con = sqlite3.connect("/var/unbound/data/unbound.sqlite")
+        self.cursor = self.con.cursor()
+        self.timer = 0
+        self.flush_interval = flush_interval
+        self.count = 0
+        self.buffer = deque()
+        self.buf_max = 4000
+
+        try:
+            self.cursor.execute("""
+                CREATE TABLE IF NOT EXISTS query (
+                    qid INTEGER PRIMARY KEY,
+                    time INTEGER,
+                    client TEXT,
+                    family TEXT,
+                    type TEXT,
+                    domain TEXT,
+                    action INTEGER,
+                    response_type INTEGER
+                )
+            """)
+            self.cursor.execute("PRAGMA journal_mode = WAL") # persists across connections
+            for size in [300, 60]:
+                self.create_bucket(size)
+            self.cursor.execute("CREATE INDEX IF NOT EXISTS idx_query ON query(time)")
+        except sqlite3.DatabaseError as e:
+            syslog.syslog(syslog.LOG_ERR, "Unable to set up database: %s" % e)
+            self.close()
+            sys.exit(1)
+
+    def create_bucket(self, interval):
+        bucket = """
+            CREATE VIEW IF NOT EXISTS v_time_buckets_{min}min AS
+            WITH RECURSIVE
+                cnt(x) AS (
+                    SELECT (unixepoch() - (unixepoch() % {intv})) x
+                    UNION ALL
+                    SELECT x - {intv} FROM cnt
+                    LIMIT {lim}
+                )
+            SELECT x as start_timestamp,
+                x + {intv} as end_timestamp
+            FROM cnt;
+        """.format(intv=interval, min=(interval//60), lim=((60//(interval//60))*24))
+        self.cursor.execute(bucket)
+
+    def rotate_db(self, interval=7):
+        t = datetime.date.today() - datetime.timedelta(days=interval)
+        query = """
+            DELETE
+            FROM query
+            WHERE DATETIME(time, 'unixepoch') < DATETIME('{time}');
+        """.format(time=t)
```

Comment:
best use bind parameters in these case, e.g. `self.cursor.execute(sql, {'prop', val})` (https://docs.python.org/3/library/sqlite3.html#sqlite3-placeholders)